staging/src/slime.io/slime/modules/meshregistry/pkg/source/zookeeper/watching.go

```diff
@@ -32,228 +35,498 @@ func (s *Source) ServiceNodeDelete(path string) {
+	// - /{root:dubbo}/{service:<service-name>}/providers/dubbo://ip:port/{provider_service}}?xxx
+	// - /{root:dubbo}/{service:<service-name>}/consumers/consumer://ip[:port]/{consumer_service}?xxx
+	servicePath string
+
+	// update by current status
+	handler func(providers, consumers []string)
+
+	serviceDeleteHandler func()
+
+	gatewatModel bool
+
+	signalExit, exit chan struct{}
+
+	initCallback func()
+}
+
+// non block
+func (ew *EndpointWatcher) Start(ctx context.Context) {
+	scope.Debugf("zk endpointWatcher %q start watching", ew.servicePath)
+	if !ew.gatewatModel {
+		go ew.watchBoth(ctx)
+	} else {
+		go ew.watchProviderOnly(ctx)
+	}
+}
+
+func (ew *EndpointWatcher) Exit() {
+	close(ew.signalExit)
+	// wait for exit
+	<-ew.exit
+}
+
+func (ew *EndpointWatcher) watchProviderOnly(ctx context.Context) {
+	defer close(ew.exit)
+
+	// Try to initialize, but it is not required to be completed,
+	// because the service may have been deleted or for others.
+	providers, _, err := ew.conn.Load().(*zk.Conn).Children(ew.servicePath + providerPathSuffix)
+	if err == nil {
+		// ignore errors when getting consumer fails
+		ew.handler(providers, nil)
 	}
-	scope.Infof("try to watch the zk path: %q", path)
-	initWatch := true
+	ew.initCallback()
+
 	for {
-		children, _, e, err := s.Con.ChildrenW(path)
-		if err != nil {
-			if shouldStopWatch(err) {
-				if !initWatch {
-					if deleteFunction != nil {
-						deleteFunction(path)
-					}
-					s.watchPath.Remove(path)
-				}
-				scope.Warnf("stop watch the zk path %q with err: %v", path, err)
-				return
-			}
-			scope.Errorf("watch the zk path %q failed: %v", path, err)
-			time.Sleep(time.Second)
+		select {
+		case <-ew.signalExit:
+			ew.serviceDeleteHandler()
+			scope.Infof("endpointWatcher %q's providerOnlyWatching exit due to service deleted", ew.servicePath)
+			return
+		default:
 		}
-		if initWatch {
-			s.watchPath.Set(path, nil)
-			scope.Infof("watching the zk path: %q", path)
-			initWatch = false
+
+		select {
+		case <-ctx.Done():
+			scope.Debugf("endpointWatcher %q's providerOnlyWatching exit due to context close", ew.servicePath)
+			return
+		default:
 		}
-		updateFunction(children, path)
-		for watchEvent := range e {
-			switch watchEvent.Type {
-			case zk.EventNodeDeleted:
-				if deleteFunction != nil {
-					deleteFunction(path)
-				}
-				s.watchPath.Remove(path)
-				scope.Infof("stop watch the zk path %q because the node was deleted", path)
-				return
-			case zk.EventResyncWatch:
-				scope.Infof("children under the zk path %q resync", path)
-				if children, _, err := s.Con.Children(path); err == nil {
-					updateFunction(children, path)
-				}
-			default:
-				// ignore other events as the channel will be closed unless the watch events are EventResyncWatch.
-				scope.Infof("children under the zk path %q maybe changed, rewatch and update it", path)
-			}
+
+		providers, _, eventCh, err := ew.conn.Load().(*zk.Conn).ChildrenW(ew.servicePath + providerPathSuffix)
```

Comment:
init `Children` and `ChildrenW` will cause dup data transfer, at least we should add a `TODO` to optimize it later

Reply:
added todo